torch/_inductor/fx_passes/comm_fusion.py

```diff
@@ -0,0 +1,154 @@
+import logging
+import operator
+from typing import Any, Callable, Optional, Union
+
+import torch
+from torch._dynamo.utils import counters
+from torch._utils_internal import print_graph
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    get_arg_value,
+    MULTIPLE,
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+
+logger = logging.getLogger(__name__)
+
+
+comm_fusion_patterns = PatternMatcherPass()
+
+
+def comm_fusion_passes(graph: torch.fx.Graph) -> None:
+    """Pattern matching passes for collective communication fucntion fusion."""
+    if not torch.distributed.is_available:
+        logger.info(
+            "Warning: torch.distributed is not available, skip communication fusion"
+        )
+        return
+
+    print_graph(graph, "Before communication fusion in post grad pass.")
+    comm_fusion_patterns.apply(graph)
+    print_graph(graph, "After communication fusion in post grad pass.")
+
+
+@register_graph_pattern(
+    CallFunctionVarArgs(torch.ops.c10d_functional.all_reduce.default, users=MULTIPLE),
+    pass_dict=comm_fusion_patterns,
+)
+def allreduce_split_to_reducescatter(match, *args, **kwargs):
+    """
+    Replace the following pattern:
+        AllReduce() - wait() - split - getitem
+
+    with:
+        ReduceScatter() - wait()
+
+    """
+    node = match.nodes[0]
+    graph = match.graph
+    if len(node.args) != 5:
+        logger.warning("AllReduce node has %s args, expected 5", len(node.args))
+        return
+
+    # Match the pattern, for now the it only works with single user/child node.
+    wait_node = CommFusionHelper.get_only_child(
+        node, "call_function", torch.ops.c10d_functional.wait_tensor.default
+    )
+    if wait_node is None:
+        return
+
+    split_node = CommFusionHelper.get_only_child(
+        wait_node, "call_function", torch.ops.aten.split.Tensor
+    )
+    if split_node is None:
+        return
+
+    getitem_node = CommFusionHelper.get_only_child(
+        split_node, "call_function", operator.getitem
+    )
+    if getitem_node is None or len(getitem_node.users) != 1:
+        return
+
+    # Make sure the output is contiguous to ensure numerical correctness
+    if not getitem_node.meta["val"].is_contiguous():
+        return
+
+    # Only support split on default dim=0, check split size
+    split_dim = get_arg_value(split_node, 2, "dim") or 0
+    if split_dim != 0:
+        return
+
+    split_size = split_node.args[1]
+    all_reduce_shape = node.meta["val"].shape[split_dim]
+    if split_size * len(node.args[3]) != all_reduce_shape:
+        return
+
+    # Make sure the getitem index is the same as the rank idx
+    current_rank = torch.distributed.get_rank()
```

Comment:
Should we actually find the pg used for the allreduce op and then use the local rank from that pg for matching and constructing the new reducescatter?

Reply:
Yeah, this is what we are doing now:

1. pg or rank_list is first extracted from the allreduce `args[3]`;
2. we then get the local rank and check if it is already in the above rank_list, and if the index match the following getitem() input;
3. after passing the check we reuse the pg from step 1 and construct the new reducescatter.

Let me add more context in the comment to make it clear.

Reply:
Hey @yoyoyocmu, we'll soon migrate to a different set of aten-level collective IRs, which replace `(tag, ranks)` with `group_name` in the dispatcher op signature (see details here: https://github.com/pytorch/pytorch/issues/93173#issuecomment-1907095208). 

It should be trivial to adapt this pass once the migration happens. Just a heads up.

Reply:
I think @wconstab has a great point: torch.distributed.get_rank() gives you the rank id of the global PG. We should probably get the PG used by the allreduce. Otherwise I'd expect this won't work? (if you have 2D parallelism and the EP is in the inner dimension)

Reply:
The `rank_list` input of `all_reduce` is actually a list of global ranks (verified with 128 GPUs in the 2D case).